File: backend/app/schemas/salary.py
from datetime import date, datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class SalaryComponentCreate(BaseModel):
    name: str = Field(..., example="Basic Salary")
    code: str = Field(..., example="BASIC")
    component_type: str = Field(..., example="Earning") # Earning, Deduction
    calculation_type: str = Field("Fixed", example="Percentage") # Fixed, Percentage, Formula
    percentage_base_code: Optional[str] = None
    percentage_value: Optional[float] = None
    formula_expression: Optional[str] = None
    frequency: str = "Monthly"
    is_taxable: bool = True
    is_statutory_applicable: bool = True
    is_prorated: bool = True
    rounding_rule: str = "Nearest"
    min_amount: Optional[float] = None
    max_amount: Optional[float] = None


class SalaryComponentResponse(SalaryComponentCreate):
    id: int

    model_config = {"from_attributes": True}


class StructureComponentItem(BaseModel):
    component_id: int
    default_amount: Optional[float] = None
    default_percentage: Optional[float] = None


class SalaryStructureCreate(BaseModel):
    company_id: int
    name: str
    description: Optional[str] = None
    components: List[StructureComponentItem] = []


class SalaryStructureResponse(BaseModel):
    id: int
    company_id: int
    name: str
    description: Optional[str] = None
    is_active: bool

    model_config = {"from_attributes": True}


class EmployeeSalaryAssign(BaseModel):
    employee_id: int
    salary_structure_id: Optional[int] = None
    total_ctc: float = Field(..., example=600000.0) # Annual CTC
    effective_date: date
    currency: str = "INR"


class EmployeeSalaryComponentResponse(BaseModel):
    id: int
    component_id: int
    monthly_amount: float
    annual_amount: float
    component: SalaryComponentResponse

    model_config = {"from_attributes": True}


class EmployeeSalaryResponse(BaseModel):
    id: int
    employee_id: int
    salary_structure_id: Optional[int] = None
    total_ctc: float
    gross_salary: float
    net_salary: float
    effective_date: date
    currency: str
    is_active: bool
    assigned_components: List[EmployeeSalaryComponentResponse] = []

    model_config = {"from_attributes": True}


class SalaryRevisionCreate(BaseModel):
    employee_id: int
    new_ctc: float
    effective_date: date
    revision_reason: Optional[str] = None


class SalaryRevisionResponse(BaseModel):
    id: int
    employee_id: int
    effective_date: date
    old_ctc: float
    new_ctc: float
    increment_percentage: float
    revision_reason: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/salary.ts
export interface SalaryComponent {
  id: number;
  name: string;
  code: string;
  component_type: string; // Earning, Deduction
  calculation_type: string; // Fixed, Percentage, Formula
  percentage_base_code?: string;
  percentage_value?: number;
  formula_expression?: string;
  frequency: string;
  is_taxable: boolean;
  is_statutory_applicable: boolean;
  is_prorated: boolean;
  rounding_rule: string;
}

export interface SalaryStructure {
  id: number;
  company_id: number;
  name: string;
  description?: string;
  is_active: boolean;
}

export interface EmployeeSalaryComponent {
  id: number;
  component_id: number;
  monthly_amount: number;
  annual_amount: number;
  component: SalaryComponent;
}

export interface EmployeeSalary {
  id: number;
  employee_id: number;
  salary_structure_id?: number;
  total_ctc: number;
  gross_salary: number;
  net_salary: number;
  effective_date: string;
  currency: string;
  is_active: boolean;
  assigned_components: EmployeeSalaryComponent[];
}

export interface SalaryRevision {
  id: number;
  employee_id: number;
  effective_date: string;
  old_ctc: number;
  new_ctc: number;
  increment_percentage: number;
  revision_reason?: string;
  created_at: string;
}
